src/hardware.rs

```diff
@@ -0,0 +1,258 @@
+            6 => Ok(Register::R6),
+            7 => Ok(Register::R7),
+            8 => Ok(Register::PC),
+            9 => Ok(Register::Cond),
+            _ => Err(VMError::Conversion(String::from(
+                "Invalid u16 for Register conversion",
+            ))),
+        }
+    }
+}
+
+/// Abstraction of the registers storage.
+pub struct Registers {
+    inner: [u16; REGS_COUNT],
+}
+
+impl Registers {
+    /// Creates a new instance of Registers with all the values of the registers
+    /// set to 0
+    pub fn new() -> Self {
+        Self {
+            inner: [0; REGS_COUNT],
+        }
+    }
+}
+
+impl Index<Register> for Registers {
+    type Output = u16;
+
+    fn index(&self, reg: Register) -> &Self::Output {
+        let index = reg.index();
+        &self.inner[index]
+    }
+}
+
+impl IndexMut<Register> for Registers {
+    fn index_mut(&mut self, reg: Register) -> &mut Self::Output {
+        let index = reg.index();
+        &mut self.inner[index]
+    }
+}
+
+/// Opcodes that identify an operation
+/// that the VM supports.
+pub enum OpCode {
+    Br,
+    Add,
+    Ld,
+    St,
+    Jsr,
+    And,
+    Ldr,
+    Str,
+    Not,
+    Ldi,
+    Sti,
+    Jmp,
+    Lea,
+    Trap,
+}
+
+impl TryFrom<u16> for OpCode {
+    type Error = VMError;
+
+    fn try_from(value: u16) -> Result<Self, Self::Error> {
+        match value {
+            0b0000 => Ok(OpCode::Br),
+            0b0001 => Ok(OpCode::Add),
+            0b0010 => Ok(OpCode::Ld),
+            0b0011 => Ok(OpCode::St),
+            0b0100 => Ok(OpCode::Jsr),
+            0b0101 => Ok(OpCode::And),
+            0b0110 => Ok(OpCode::Ldr),
+            0b0111 => Ok(OpCode::Str),
+            0b1001 => Ok(OpCode::Not),
+            0b1010 => Ok(OpCode::Ldi),
+            0b1011 => Ok(OpCode::Sti),
+            0b1100 => Ok(OpCode::Jmp),
+            0b1110 => Ok(OpCode::Lea),
+            0b1111 => Ok(OpCode::Trap),
+            _ => Err(VMError::Conversion(String::from(
+                "Invalid u16 for OpcCode conversion",
+            ))),
+        }
+    }
+}
+
+/// Condition flags that indicate
+/// the result of the previous calculation
+pub enum CondFlag {
+    Pos = 1 << 0,
+    Zro = 1 << 1,
+    Neg = 1 << 2,
+}
+
+impl CondFlag {
+    pub fn value(&self) -> u16 {
+        match self {
+            CondFlag::Pos => 1 << 0,
+            CondFlag::Zro => 1 << 1,
+            CondFlag::Neg => 1 << 2,
+        }
+    }
+}
+
+/// Registers that are located on the memory
+#[derive(Clone, Copy)]
+pub enum MemoryRegister {
+    KeyboardStatus,
+    KeyboardData,
+}
+
+impl MemoryRegister {
+    fn address(&self) -> u16 {
+        match self {
+            MemoryRegister::KeyboardStatus => 0xFE00,
+            MemoryRegister::KeyboardData => 0xFE02,
+        }
+    }
+}
+
+impl From<MemoryRegister> for usize {
+    fn from(value: MemoryRegister) -> Self {
+        value.address().into()
+    }
+}
+
+/// Allows us to compare an u16 with a MemoryRegister
```

Comment:
To keep consistency with the documentation style, I would remove the `us`:

```suggestion
/// Allows comparison of an u16 with a MemoryRegister
```